fairscale/experimental/nn/multiprocess_pipe.py

```diff
@@ -85,122 +42,629 @@ def dloss(input_rref: rpc.RRef, target_rref: rpc.RRef) -> rpc.RRef:
+            s0 = current_stream(result.device)
+            if is_cuda(s0):
+                # TODO. Investigate why this is needed and remove it if possible.
+                as_cuda(s0).synchronize()
+            return result
+
+        return None
+
+
+class MultiInputSequential(nn.Module):
+    """A variation of nn.Sequential, that allows the first module in the sequence accepts
+        multiple inputs. To be used internally by _split_module
+    """
+
+    def __init__(self, *modules: nn.Module) -> None:
+        super().__init__()
+        self.modules_list = modules
+
+    def forward(self, *inputs: Tuple[Tensor]) -> Tensor:  # type: ignore
+        input = self.modules_list[0](*inputs)
+        for module in self.modules_list[1:]:
+            input = module(input)
+        return input
+
 
-        model = nn.Sequential(a, b, c, d)
-        model = Pipe(model, balance=[1, 1, 1, 1], chunks=8)
-        output = model(input)
+def RemoteSequential(rref_list: List[rpc.RRef]) -> MultiInputSequential:
+    return MultiInputSequential(*(r.local_value() for r in rref_list))
 
-    .. _Pipe: https://arxiv.org/abs/1811.06965
 
-    Pipe combines pipeline parallelism with checkpointing to reduce peak
+def _split_module(graph: PipelineModulesGraph) -> List[Tuple[List[int], rpc.RRef]]:
+    """Splits the graph into pipeline partitions and for each parition returns a tuple (indices, module_rref),
+    where indices is indices of modules of the partition in the graph, and module_rref is an RRef to an nn.Module:
+    If there is only one module in the partition, module_rref is reference to that module; otherwise those modules
+    are wrapped by a MultiInputSequential and module_rref referes to that.
+    """
+    graph.compile_output_users()
+    module_used = [False] * len(graph.modules_list)
+    partitions = []
+    for module_idx, module in enumerate(graph.modules_list):
+        if module_used[module_idx]:
+            continue
+        partition = []
+        current_module_idx = module_idx
+        current_module = module
+        while True:
+            assert not module_used[current_module_idx]
+            module_used[current_module_idx] = True
+            partition.append(current_module_idx)
+            # If we reached a module with multiple outputs or with multiple users for its output,
+            # stop adding more modules to the partition.
+            if len(graph.output_users[current_module_idx]) != 1:
+                break
+            if graph.modules_list[current_module_idx].num_outputs is not None:
+                break
+            # Next module to add is the only consumer of the ouput of the current module
+            next_module_idx = graph.output_users[current_module_idx][0][0]
+            next_module = graph.modules_list[next_module_idx]
+            # If the next module has multiple inputs, do not add it to the current partition and stop.
+            if graph.inputs[next_module_idx] != [(current_module_idx, 0)]:
```

Comment:
It would be nice for knowledge of graph.inputs to not be exposed outside the graph class.